src/main/java/net/minecraftforge/unification/UnificationManager.java

```diff
@@ -0,0 +1,147 @@
+     * @param outputCopier copies values so they can't be mutated. If T is immutable, just return it without copying.
+     */
+    public static <T> UnificationRegistry<T> createRegistry(Function<T, ResourceLocation> namer, Function<T, T> validator, Function<T, T> outputCopier)
+    {
+        return new UnificationRegistry<>(namer, validator, outputCopier);
+    }
+
+    /**
+     * @return all the {@link UnificationMaterial} that have been added.
+     */
+    public static Collection<UnificationMaterial> getMaterials()
+    {
+        return Collections.unmodifiableCollection(MATERIALS.values());
+    }
+
+    /**
+     * @return all the {@link UnificationForm} that have been added.
+     */
+    public static Collection<UnificationForm> getForms()
+    {
+        return Collections.unmodifiableCollection(FORMS.values());
+    }
+
+    /**
+     * Get an {@link UnificationMaterial} from a unique Id. Creates one if it does not exist.
+     * Many are already defined in {@link UnificationConstants}.
+     */
+    public static UnificationMaterial getMaterial(String materialUid)
+    {
+        Preconditions.checkNotNull(materialUid, "uid must not be null.");
+        return MATERIALS.computeIfAbsent(materialUid, UnificationMaterial::new);
+    }
+
+    /**
+     * Get an {@link UnificationForm} from a unique Id. Creates one if it does not exist.
+     * Many are already defined in {@link UnificationConstants}.
+     */
+    public static UnificationForm getForm(String typeUid)
+    {
+        Preconditions.checkNotNull(typeUid, "uid must not be null.");
+        return FORMS.computeIfAbsent(typeUid, UnificationForm::new);
+    }
+
+    private static UnificationRegistry<ItemStack> createItemStackUnificationRegistry()
+    {
+        Function<ItemStack, ResourceLocation> namer = itemStack -> {
+            ResourceLocation registryName = itemStack.getItem().getRegistryName();
+            Preconditions.checkNotNull(registryName, "registry name must not be null");
+            return registryName;
+        };
+        Function<ItemStack, ItemStack> validator = itemStack -> {
+            Preconditions.checkNotNull(itemStack, "itemStack must not be null");
+            Preconditions.checkArgument(!itemStack.isEmpty(), "itemStack must not be empty");
+            Preconditions.checkArgument(itemStack.getMetadata() != OreDictionary.WILDCARD_VALUE, "Product must be an actual ItemStack, not a wildcard.");
+            itemStack = itemStack.copy();
+            itemStack.setCount(1);
+            return itemStack;
+        };
+        return createRegistry(namer, validator, ItemStack::copy);
+    }
+
+    private static UnificationRegistry<IBlockState> createBlockStateUnificationRegistry()
+    {
+        Function<IBlockState, ResourceLocation> namer = blockState -> {
+            ResourceLocation registryName = blockState.getBlock().getRegistryName();
+            Preconditions.checkNotNull(registryName, "registry name must not be null");
```

Comment:
`return Preconditions.checkNotNull(registryName, "registry name must not be null");`

Reply:
Ehh I think it's fine the way it is, that's more a style choice than anything.